gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/VectoredIOImpl.java

```diff
@@ -0,0 +1,289 @@
+      ByteBuffer dst = allocate.apply(combinedFileRange.getLength());
+      int numRead = channel.read(dst.duplicate());
+      logger.atFiner().log(
+          "Read combinedFileRange completed from range: %s, path: %s, readBytes: %d",
+          combinedFileRange, gcsPath, numRead);
+      populateChildBuffer(combinedFileRange, dst, numRead);
+    } catch (Exception e) {
+      logger.atWarning().withCause(e).log(
+          "Exception while reading combinedFileRange:%s for path: %s", combinedFileRange, gcsPath);
+      combinedFileRange.getData().completeExceptionally(e);
+      // complete exception all the underlying ranges which have not already
+      // finished.
+      for (FileRange child : combinedFileRange.getUnderlying()) {
+        if (!child.getData().isDone()) {
+          child
+              .getData()
+              .completeExceptionally(
+                  new IOException(
+                      String.format(
+                          "Error while populating childRange: %s from combinedRange: %s for path: %s",
+                          child, combinedFileRange, gcsPath),
+                      e));
+        }
+      }
+    }
+  }
+
+  /**
+   * Populate the child ranges from the ByteBuffer and update the combinedFileRange completion
+   * status. If readBytes < requested Populate the child ranges which can be served from
+   * partialContent and throw afterwards.
+   *
+   * @param combinedFileRange
+   * @param readContent
+   * @param numRead size of read content
+   * @throws EOFException in case partial content is read.
+   */
+  private void populateChildBuffer(
+      CombinedFileRange combinedFileRange, ByteBuffer readContent, int numRead)
+      throws EOFException {
+    if (numRead < 0) {
+      throw new EOFException(
+          String.format(
+              "EOF reached before whole combinedFileRange can be read, range: %s, path: %s",
+              combinedFileRange, gcsPath));
+    }
+    // content was read partially
+    // can happen when request range is beyond file size
+    if (numRead < combinedFileRange.getLength()) {
+      int readBytesCumulative = 0;
+      // populating all child ranges for which we have content
+      for (FileRange child : combinedFileRange.getUnderlying()) {
+        readBytesCumulative += child.getLength();
+        if (readBytesCumulative <= numRead) {
+          ByteBuffer childBuffer = sliceTo(readContent, combinedFileRange.getOffset(), child);
+          child.getData().complete(childBuffer);
+        } else {
+          // remaining child ranges needs be marked appropriately in caller.
+          throw new EOFException(
+              String.format(
+                  "EOF reached before whole range can be read, combinedFileRange: %s, expected length: %s, readBytes: %s, path: %s",
+                  combinedFileRange, combinedFileRange.getLength(), numRead, gcsPath));
+        }
+      }
+    }
+
+    for (FileRange child : combinedFileRange.getUnderlying()) {
```

Comment:
By twice you mean for loop at L:169 and here?

I guess comment before L:166 is to explain just that. When we readcontent in combined range. It can happen taht we have got less data that requested. For e.g. fileSize = 10, mergedRnageRequest:[offset:5,length:15]. In this case we will get partial data of length 5 from underneath channel.

Reply:
It is not clear what we want to achieve here.

In line 173 we mark it as complete. We are doing the same in line 186. The caller might not see the effect of line 185, correct?

Reply:
L:173 will only be effective if content is partially read, where we mark some of the ranges with contetn and other with exception.

L:185 will only be effective if content was fully read.